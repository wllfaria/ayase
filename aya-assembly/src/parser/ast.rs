use std::ops::Range;

use aya_cpu::op_code::OpCode;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum Operator {
    Add,
    Sub,
    Mul,
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
pub struct ByteOffset {
    pub start: usize,
    pub end: usize,
}

impl From<ByteOffset> for Range<usize> {
    fn from(offset: ByteOffset) -> Range<usize> {
        offset.start..offset.end
    }
}

impl From<ByteOffset> for miette::SourceSpan {
    fn from(value: ByteOffset) -> Self {
        Self::new(value.start.into(), value.end - value.start)
    }
}

impl From<Range<usize>> for ByteOffset {
    fn from(value: Range<usize>) -> Self {
        Self {
            start: value.start,
            end: value.end,
        }
    }
}

#[derive(Debug)]
pub struct Ast {
    pub statements: Vec<Statement>,
}

#[derive(Debug, PartialEq, Eq)]
pub enum Statement {
    Instruction(Box<Instruction>),
    HexLiteral(ByteOffset),
    Address(ByteOffset),
    Register(ByteOffset),
    Var(ByteOffset),
    Operator(Operator),
    Label {
        name: ByteOffset,
        exported: bool,
    },
    FieldAccessor {
        module: ByteOffset,
        field: ByteOffset,
    },
    Import {
        name: ByteOffset,
        path: ByteOffset,
        address: ByteOffset,
        variables: Vec<Statement>,
    },
    ImportVar {
        name: ByteOffset,
        value: Box<Statement>,
    },
    Data {
        name: ByteOffset,
        size: u8,
        exported: bool,
        values: Vec<Statement>,
    },
    Const {
        name: ByteOffset,
        exported: bool,
        value: Box<Statement>,
    },
    BinaryOp {
        lhs: Box<Statement>,
        operator: Operator,
        rhs: Box<Statement>,
    },
}

impl From<Instruction> for Statement {
    fn from(instruction: Instruction) -> Self {
        Self::Instruction(Box::new(instruction))
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum InstructionKind {
    LitReg,
    RegReg,
    RegMem,
    MemReg,
    LitMem,
    RegPtrReg,
    NoArgs,
    SingleReg,
    SingleLit,
}

impl InstructionKind {
    pub fn byte_size(&self) -> u8 {
        match self {
            InstructionKind::LitReg => 4,
            InstructionKind::RegReg => 3,
            InstructionKind::RegMem => 4,
            InstructionKind::MemReg => 4,
            InstructionKind::LitMem => 5,
            InstructionKind::RegPtrReg => 3,
            InstructionKind::NoArgs => 1,
            InstructionKind::SingleReg => 2,
            InstructionKind::SingleLit => 3,
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum Instruction {
    MovLitReg(Statement, Statement),
    MovRegReg(Statement, Statement),
    MovRegMem(Statement, Statement),
    MovMemReg(Statement, Statement),
    MovLitMem(Statement, Statement),
    MovRegPtrReg(Statement, Statement),
    AddRegReg(Statement, Statement),
    AddLitReg(Statement, Statement),
    SubRegReg(Statement, Statement),
    SubLitReg(Statement, Statement),
    MulRegReg(Statement, Statement),
    MulLitReg(Statement, Statement),
    LshRegReg(Statement, Statement),
    LshLitReg(Statement, Statement),
    RshRegReg(Statement, Statement),
    RshLitReg(Statement, Statement),
    AndRegReg(Statement, Statement),
    AndLitReg(Statement, Statement),
    OrLitReg(Statement, Statement),
    OrRegReg(Statement, Statement),
    XorLitReg(Statement, Statement),
    XorRegReg(Statement, Statement),
    Inc(Statement),
    Dec(Statement),
    Not(Statement),
    JeqLit(Statement, Statement),
    JeqReg(Statement, Statement),
    JgtLit(Statement, Statement),
    JgtReg(Statement, Statement),
    JneLit(Statement, Statement),
    JneReg(Statement, Statement),
    JgeLit(Statement, Statement),
    JgeReg(Statement, Statement),
    JleLit(Statement, Statement),
    JleReg(Statement, Statement),
    JltLit(Statement, Statement),
    JltReg(Statement, Statement),
    Jmp(Statement),
    PshLit(Statement),
    PshReg(Statement),
    Pop(Statement),
    CallLit(Statement),
    Ret,
    Hlt,
}

impl Instruction {
    pub fn lhs(&self) -> &Statement {
        match self {
            Instruction::MovLitReg(lhs, _)
            | Instruction::MovRegReg(lhs, _)
            | Instruction::MovRegMem(lhs, _)
            | Instruction::MovMemReg(lhs, _)
            | Instruction::MovLitMem(lhs, _)
            | Instruction::MovRegPtrReg(lhs, _)
            | Instruction::AddRegReg(lhs, _)
            | Instruction::AddLitReg(lhs, _)
            | Instruction::SubRegReg(lhs, _)
            | Instruction::SubLitReg(lhs, _)
            | Instruction::MulRegReg(lhs, _)
            | Instruction::MulLitReg(lhs, _)
            | Instruction::LshRegReg(lhs, _)
            | Instruction::LshLitReg(lhs, _)
            | Instruction::RshRegReg(lhs, _)
            | Instruction::RshLitReg(lhs, _)
            | Instruction::AndRegReg(lhs, _)
            | Instruction::AndLitReg(lhs, _)
            | Instruction::OrLitReg(lhs, _)
            | Instruction::OrRegReg(lhs, _)
            | Instruction::XorLitReg(lhs, _)
            | Instruction::XorRegReg(lhs, _)
            | Instruction::JeqLit(lhs, _)
            | Instruction::JeqReg(lhs, _)
            | Instruction::JgtLit(lhs, _)
            | Instruction::JgtReg(lhs, _)
            | Instruction::JneLit(lhs, _)
            | Instruction::JneReg(lhs, _)
            | Instruction::JgeLit(lhs, _)
            | Instruction::JgeReg(lhs, _)
            | Instruction::JleLit(lhs, _)
            | Instruction::JleReg(lhs, _)
            | Instruction::JltLit(lhs, _)
            | Instruction::JltReg(lhs, _)
            | Instruction::PshLit(lhs)
            | Instruction::PshReg(lhs)
            | Instruction::Pop(lhs)
            | Instruction::CallLit(lhs)
            | Instruction::Inc(lhs)
            | Instruction::Dec(lhs)
            | Instruction::Jmp(lhs)
            | Instruction::Not(lhs) => lhs,

            Instruction::Ret | Instruction::Hlt => unreachable!(),
        }
    }

    pub fn rhs(&self) -> &Statement {
        match self {
            Instruction::MovLitReg(_, rhs)
            | Instruction::MovRegReg(_, rhs)
            | Instruction::MovRegMem(_, rhs)
            | Instruction::MovMemReg(_, rhs)
            | Instruction::MovLitMem(_, rhs)
            | Instruction::MovRegPtrReg(_, rhs)
            | Instruction::AddRegReg(_, rhs)
            | Instruction::AddLitReg(_, rhs)
            | Instruction::SubRegReg(_, rhs)
            | Instruction::SubLitReg(_, rhs)
            | Instruction::MulRegReg(_, rhs)
            | Instruction::MulLitReg(_, rhs)
            | Instruction::LshRegReg(_, rhs)
            | Instruction::LshLitReg(_, rhs)
            | Instruction::RshRegReg(_, rhs)
            | Instruction::RshLitReg(_, rhs)
            | Instruction::AndRegReg(_, rhs)
            | Instruction::AndLitReg(_, rhs)
            | Instruction::OrLitReg(_, rhs)
            | Instruction::OrRegReg(_, rhs)
            | Instruction::XorLitReg(_, rhs)
            | Instruction::XorRegReg(_, rhs)
            | Instruction::JeqLit(_, rhs)
            | Instruction::JeqReg(_, rhs)
            | Instruction::JgtLit(_, rhs)
            | Instruction::JgtReg(_, rhs)
            | Instruction::JneLit(_, rhs)
            | Instruction::JneReg(_, rhs)
            | Instruction::JgeLit(_, rhs)
            | Instruction::JgeReg(_, rhs)
            | Instruction::JleLit(_, rhs)
            | Instruction::JleReg(_, rhs)
            | Instruction::JltLit(_, rhs)
            | Instruction::JltReg(_, rhs) => rhs,

            Instruction::PshLit(_)
            | Instruction::PshReg(_)
            | Instruction::Pop(_)
            | Instruction::CallLit(_)
            | Instruction::Inc(_)
            | Instruction::Dec(_)
            | Instruction::Not(_)
            | Instruction::Jmp(_)
            | Instruction::Ret
            | Instruction::Hlt => unreachable!(),
        }
    }

    pub fn opcode(&self) -> OpCode {
        match self {
            Instruction::MovLitReg(_, _) => OpCode::MovLitReg,
            Instruction::MovRegReg(_, _) => OpCode::MovRegReg,
            Instruction::MovRegMem(_, _) => OpCode::MovRegMem,
            Instruction::MovMemReg(_, _) => OpCode::MovMemReg,
            Instruction::MovLitMem(_, _) => OpCode::MovLitMem,
            Instruction::MovRegPtrReg(_, _) => OpCode::MovRegPtrReg,

            Instruction::AddRegReg(_, _) => OpCode::AddRegReg,
            Instruction::AddLitReg(_, _) => OpCode::AddLitReg,
            Instruction::SubRegReg(_, _) => OpCode::SubRegReg,
            Instruction::SubLitReg(_, _) => OpCode::SubLitReg,
            Instruction::Inc(_) => OpCode::IncReg,
            Instruction::Dec(_) => OpCode::DecReg,
            Instruction::MulLitReg(_, _) => OpCode::MulLitReg,
            Instruction::MulRegReg(_, _) => OpCode::MulRegReg,

            Instruction::LshLitReg(_, _) => OpCode::LshLitReg,
            Instruction::LshRegReg(_, _) => OpCode::LshRegReg,
            Instruction::RshLitReg(_, _) => OpCode::RshLitReg,
            Instruction::RshRegReg(_, _) => OpCode::RshRegReg,
            Instruction::AndLitReg(_, _) => OpCode::AndLitReg,
            Instruction::AndRegReg(_, _) => OpCode::AndRegReg,
            Instruction::OrLitReg(_, _) => OpCode::OrLitReg,
            Instruction::OrRegReg(_, _) => OpCode::OrRegReg,
            Instruction::XorLitReg(_, _) => OpCode::XorLitReg,
            Instruction::XorRegReg(_, _) => OpCode::XorRegReg,
            Instruction::Not(_) => OpCode::Not,

            Instruction::PshLit(_) => OpCode::PushLit,
            Instruction::PshReg(_) => OpCode::PushReg,
            Instruction::Pop(_) => OpCode::Pop,
            Instruction::CallLit(_) => OpCode::Call,
            Instruction::Ret => OpCode::Ret,
            Instruction::Hlt => OpCode::Halt,

            Instruction::JeqLit(_, _) => OpCode::JeqLit,
            Instruction::JeqReg(_, _) => OpCode::JeqReg,
            Instruction::JgtLit(_, _) => OpCode::JgtLit,
            Instruction::JgtReg(_, _) => OpCode::JgtReg,
            Instruction::JneLit(_, _) => OpCode::JneLit,
            Instruction::JneReg(_, _) => OpCode::JneReg,
            Instruction::JgeLit(_, _) => OpCode::JgeLit,
            Instruction::JgeReg(_, _) => OpCode::JgeReg,
            Instruction::JleLit(_, _) => OpCode::JleLit,
            Instruction::JleReg(_, _) => OpCode::JleReg,
            Instruction::JltLit(_, _) => OpCode::JltLit,
            Instruction::JltReg(_, _) => OpCode::JltReg,
            Instruction::Jmp(_) => OpCode::Jmp,
        }
    }

    pub fn kind(&self) -> InstructionKind {
        match self {
            Instruction::MovLitReg(_, _)
            | Instruction::AddLitReg(_, _)
            | Instruction::SubLitReg(_, _)
            | Instruction::MulLitReg(_, _)
            | Instruction::AndLitReg(_, _)
            | Instruction::OrLitReg(_, _)
            | Instruction::LshLitReg(_, _)
            | Instruction::RshLitReg(_, _)
            | Instruction::XorLitReg(_, _) => InstructionKind::LitReg,

            Instruction::MovRegReg(_, _)
            | Instruction::AddRegReg(_, _)
            | Instruction::SubRegReg(_, _)
            | Instruction::MulRegReg(_, _)
            | Instruction::AndRegReg(_, _)
            | Instruction::OrRegReg(_, _)
            | Instruction::LshRegReg(_, _)
            | Instruction::RshRegReg(_, _)
            | Instruction::XorRegReg(_, _) => InstructionKind::RegReg,

            Instruction::MovLitMem(_, _)
            | Instruction::JneLit(_, _)
            | Instruction::JeqLit(_, _)
            | Instruction::JgtLit(_, _)
            | Instruction::JgeLit(_, _)
            | Instruction::JleLit(_, _)
            | Instruction::JltLit(_, _) => InstructionKind::LitMem,

            Instruction::Inc(_)
            | Instruction::Dec(_)
            | Instruction::Not(_)
            | Instruction::PshReg(_)
            | Instruction::Pop(_) => InstructionKind::SingleReg,

            Instruction::MovRegMem(_, _)
            | Instruction::JneReg(_, _)
            | Instruction::JeqReg(_, _)
            | Instruction::JgtReg(_, _)
            | Instruction::JgeReg(_, _)
            | Instruction::JleReg(_, _)
            | Instruction::JltReg(_, _) => InstructionKind::RegMem,

            Instruction::MovMemReg(_, _) => InstructionKind::MemReg,
            Instruction::MovRegPtrReg(_, _) => InstructionKind::RegPtrReg,
            Instruction::PshLit(_) | Instruction::CallLit(_) | Instruction::Jmp(_) => InstructionKind::SingleLit,
            Instruction::Ret | Instruction::Hlt => InstructionKind::NoArgs,
        }
    }
}
