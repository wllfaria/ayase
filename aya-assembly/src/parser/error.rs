pub type Result<T> = std::result::Result<T, miette::Error>;

pub static ADDRESS_HELP: &str = "valid addresses takes the form of &FFFF";
pub static ADDRESS_MSG: &str = "[SYNTAX_ERROR]: expected address";

pub static HEX_LIT_HELP: &str = "valid hex literals takes the form of $FFFF";
pub static HEX_LIT_MSG: &str = "[SYNTAX_ERROR]: expected hex literal";

pub static VAR_MSG: &str = "[SYNTAX_ERROR]: variable name must be a valid identifier";
pub static VAR_HELP: &str = "variables must start with a ! [BANG] followed by a valid identifier";

pub static REGISTER_MSG: &str = "[SYNTAX_ERROR]: invalid register name";
pub static REGISTER_HELP: &str = "register name must be in the set of valid registers";

pub static BRACKETED_EXPR_HELP: &str = "invalid bracketed expression";
pub static BRACKETED_EXPR_MSG: &str = "[SYNTAX_ERROR]: invalid bracketed expression";

pub static UNTERMINATED_STRING_HELP: &str = "did you forget a closing \"";
pub static UNTERMINATED_STRING_MSG: &str = "unterminated string";

pub static PATH_MSG: &str = "[SYNTAX_ERROR]: expected path string";

pub static IDENT_MSG: &str = "[SYNTAX_ERROR]: expected valid identifier";

pub static EOF_MSG: &str = "[SYNTAX_ERROR]: unexpected end of file [EOF]";

pub static UNEXPECTED_TOKEN_MSG: &str = "[SYNTAX_ERROR]: unexpected token";

pub static COLON_MSG: &str = "[SYNTAX_ERROR]: expected a `:` [COLON]";
pub static DOT_MSG: &str = "[SYNTAX_ERROR]: expected a `.` [DOT]";
pub static PLUS_MSG: &str = "[SYNTAX_ERROR]: expected a `+` [PLUS]";
pub static COMMA_MSG: &str = "[SYNTAX_ERROR]: expected a `,` [COMMA]";
pub static LBRACE_MSG: &str = "[SYNTAX_ERROR]: expected a `{` [LEFT_CURLY]";
pub static RBRACE_MSG: &str = "[SYNTAX_ERROR]: expected a `}` [RIGHT_CURLY]";
pub static LBRACKET_MSG: &str = "[SYNTAX_ERROR]: expected a `[` [LEFT_BRACKET]";
pub static RBRACKET_MSG: &str = "[SYNTAX_ERROR]: expected a `]` [RIGHT_BRACKET]";
