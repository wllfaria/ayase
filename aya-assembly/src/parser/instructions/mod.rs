mod add;
mod and;
mod call;
mod dec;
mod hlt;
mod inc;
mod jeq;
mod jge;
mod jgt;
mod jle;
mod jlt;
mod jmp;
mod jne;
mod lsh;
mod mov;
mod mul;
mod not;
mod or;
mod pop;
mod psh;
mod ret;
mod rsh;
mod sub;
mod xor;

pub use add::parse_add;
pub use and::parse_and;
pub use call::parse_call;
pub use dec::parse_dec;
pub use hlt::parse_hlt;
pub use inc::parse_inc;
pub use jeq::parse_jeq;
pub use jge::parse_jge;
pub use jgt::parse_jgt;
pub use jle::parse_jle;
pub use jlt::parse_jlt;
pub use jmp::parse_jmp;
pub use jne::parse_jne;
pub use lsh::parse_lsh;
pub use mov::parse_mov;
pub use mul::parse_mul;
pub use not::parse_not;
pub use or::parse_or;
pub use pop::parse_pop;
pub use psh::parse_psh;
pub use ret::parse_ret;
pub use rsh::parse_rsh;
pub use sub::parse_sub;
pub use xor::parse_xor;
