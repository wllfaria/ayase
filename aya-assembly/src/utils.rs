use crate::lexer::Token;
use crate::parser::error::{EOF_MSG, UNEXPECTED_TOKEN_MSG};

pub fn bail<S: AsRef<str>>(source: S, help: S, message: S, offset: impl Into<miette::SourceSpan>) -> miette::Error {
    miette::Error::from(
        miette::MietteDiagnostic::new(message.as_ref())
            .with_labels(vec![miette::LabeledSpan::at(offset, "this bit")])
            .with_help(help.as_ref()),
    )
    .with_source_code(source.as_ref().to_string())
}

pub fn bail_multi<S: AsRef<str>>(
    source: &str,
    labels: impl IntoIterator<Item = miette::LabeledSpan>,
    message: S,
    help: S,
) -> miette::Error {
    miette::Error::from(
        miette::MietteDiagnostic::new(message.as_ref())
            .with_labels(labels)
            .with_help(help.as_ref()),
    )
    .with_source_code(source.to_string())
}

pub fn unexpected_eof<S: AsRef<str>, T>(source: S, help: S) -> miette::Result<T> {
    let end = source.as_ref().len();
    let start = end.saturating_sub(1);
    Err(bail(source.as_ref(), help.as_ref(), EOF_MSG, start..end))
}

pub fn unexpected_token<S: AsRef<str>, T>(source: S, token: &Token) -> miette::Result<T> {
    Err(bail(
        source.as_ref(),
        &format!("unexpected token {token}"),
        UNEXPECTED_TOKEN_MSG,
        token.offset(),
    ))
}

pub fn unexpected_statement<S: AsRef<str>, T>(
    source: S,
    help: S,
    offset: impl Into<miette::SourceSpan>,
) -> miette::Result<T> {
    Err(bail(
        source.as_ref(),
        help.as_ref(),
        "[SYNTAX_ERROR]: unexpected statement",
        offset,
    ))
}
