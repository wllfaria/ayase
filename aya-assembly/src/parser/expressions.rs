use miette::Result;

use super::ast::{Operator, Statement};
use super::common::{expect, parse_hex_lit, parse_register, parse_variable, peek};
use super::error::{HEX_LIT_HELP, HEX_LIT_MSG};
use crate::lexer::{Kind, Lexer};
use crate::utils::unexpected_token;

mod precedences {
    use miette::Result;

    use crate::parser::ast::Operator;

    pub const BASE: u8 = 0;
    pub const ADD: u8 = 1;
    pub const MUL: u8 = 2;

    pub fn from_operator(operator: Operator) -> Result<u8> {
        match operator {
            Operator::Add => Ok(ADD),
            Operator::Sub => Ok(ADD),
            Operator::Mul => Ok(MUL),
        }
    }
}

pub fn parse_literal_expr<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<Statement> {
    expect(Kind::LBracket, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    let value = parse_expr(source.as_ref(), lexer, precedences::BASE)?;
    expect(Kind::RBracket, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    Ok(value)
}

pub fn parse_address_expr<S: AsRef<str>>(source: S, lexer: &mut Lexer, help: S, message: S) -> Result<Statement> {
    expect(Kind::Ampersand, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    expect(Kind::LBracket, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;

    let value = parse_expr(source.as_ref(), lexer, precedences::BASE)?;

    expect(Kind::RBracket, lexer, source.as_ref(), help.as_ref(), message.as_ref())?;
    Ok(Statement::Address(Box::new(value)))
}

fn parse_expr<S: AsRef<str>>(source: S, lexer: &mut Lexer, precedence: u8) -> Result<Statement> {
    let token = peek(source.as_ref(), lexer)?;
    let mut lhs = match token.kind {
        Kind::LParen => {
            lexer.next().transpose()?;
            let value = parse_expr(source.as_ref(), lexer, precedences::BASE)?;
            expect(
                Kind::RParen,
                lexer,
                source.as_ref(),
                "you likely forgot a closing parenthesis `)` [RIGHT_PAREN]",
                "[SYNTAX_ERROR]: unterminated expression group",
            )?;
            value
        }
        Kind::HexNumber => Statement::HexLiteral(parse_hex_lit(source.as_ref(), lexer, HEX_LIT_HELP, HEX_LIT_MSG)?),
        Kind::Ident => Statement::Register(parse_register(source.as_ref(), lexer)?),
        Kind::Bang => Statement::Var(parse_variable(
            source.as_ref(),
            lexer,
            "variable name must be a valid identifier",
            "[SYNTAX_ERROR]: invalid variable name",
        )?),
        _ => unexpected_token(source.as_ref(), &token)?,
    };

    loop {
        let token = peek(source.as_ref(), lexer)?;
        match token.kind {
            Kind::RParen => break,
            Kind::RBracket => break,
            kind if !kind.is_operator() => unexpected_token(source.as_ref(), &token)?,
            _ => {}
        }

        let operator = peek(source.as_ref(), lexer)?;
        let operator = Operator::try_from(operator)?;
        let operator_precedence = precedences::from_operator(operator)?;

        if operator_precedence < precedence {
            break;
        }

        lexer.next().transpose()?;

        let rhs = parse_expr(source.as_ref(), lexer, operator_precedence)?;
        lhs = Statement::BinaryOp {
            lhs: Box::new(lhs),
            operator,
            rhs: Box::new(rhs),
        }
    }

    Ok(lhs)
}
