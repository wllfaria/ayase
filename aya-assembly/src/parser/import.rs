use super::common::{expect, parse_hex_lit, parse_identifier, parse_string, parse_variable};
use super::error::{
    Result, ADDRESS_HELP, ADDRESS_MSG, COLON_MSG, COMMA_MSG, DOT_MSG, HEX_LIT_HELP, HEX_LIT_MSG, IDENT_MSG, LBRACE_MSG,
    LBRACKET_MSG, PATH_MSG, RBRACE_MSG, RBRACKET_MSG,
};
use crate::lexer::{Kind, Lexer, TransposeRef};
use crate::parser::ast::Statement;
use crate::parser::syntax::parse_simple_address;
use crate::utils::{unexpected_eof, unexpected_token};

fn parse_field_accessor<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    let module = parse_identifier(
        source.as_ref(),
        lexer,
        "module name in import value must be a valid identifier",
        IDENT_MSG,
    )?;

    expect(
        Kind::Dot,
        lexer,
        source.as_ref(),
        "module field accessor must be dot separated",
        DOT_MSG,
    )?;

    let field = parse_identifier(
        source.as_ref(),
        lexer,
        "module name in import value must be a valid identifier",
        IDENT_MSG,
    )?;

    Ok(Statement::FieldAccessor { module, field })
}

fn parse_bracketed_value<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    expect(
        Kind::LBracket,
        lexer,
        source.as_ref(),
        "bracketed import value must start with a left bracket",
        LBRACKET_MSG,
    )?;

    let Ok(Some(token)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };

    let statement = match token.kind {
        Kind::Ident => parse_field_accessor(source.as_ref(), lexer)?,
        Kind::Bang => Statement::Var(parse_variable(
            source.as_ref(),
            lexer,
            "variable value must be a valid identifier",
            IDENT_MSG,
        )?),
        _ => return unexpected_token(source.as_ref(), token),
    };

    expect(
        Kind::RBracket,
        lexer,
        source.as_ref(),
        "unfinished bracketed import value",
        RBRACKET_MSG,
    )?;

    Ok(statement)
}

fn parse_import_var_value<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    let Ok(Some(token)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };

    match token.kind {
        Kind::LBracket => parse_bracketed_value(source.as_ref(), lexer),
        Kind::Bang => Ok(Statement::Var(parse_variable(
            source.as_ref(),
            lexer,
            "variable value must be a valid identifier",
            IDENT_MSG,
        )?)),
        Kind::HexNumber => Ok(Statement::HexLiteral(parse_hex_lit(
            source.as_ref(),
            lexer,
            HEX_LIT_HELP,
            HEX_LIT_MSG,
        )?)),
        _ => return unexpected_token(source.as_ref(), token),
    }
}

fn parse_import_vars<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Vec<Statement>> {
    let mut variables = vec![];

    loop {
        let Ok(Some(token)) = lexer.peek().transpose() else {
            let Err(err) = lexer.next().transpose() else {
                return unexpected_eof(source.as_ref(), "unterminated import statement");
            };
            return Err(err);
        };

        if token.kind == Kind::RBrace {
            break;
        }

        let name = parse_identifier(
            source.as_ref(),
            lexer,
            "variable name must be a valid identifier",
            IDENT_MSG,
        )?;

        expect(
            Kind::Colon,
            lexer,
            source.as_ref(),
            "import variable name and value must be separated by a colon",
            COLON_MSG,
        )?;

        let value = parse_import_var_value(source.as_ref(), lexer)?;

        let Ok(Some(next)) = lexer.peek().transpose() else {
            let Err(err) = lexer.next().transpose() else {
                return unexpected_eof(source.as_ref(), "unterminated import statement");
            };
            return Err(err);
        };

        match next.kind {
            Kind::RBrace => {}
            _ => {
                _ = expect(
                    Kind::Comma,
                    lexer,
                    source.as_ref(),
                    "import variables must be separated by a comma",
                    COMMA_MSG,
                )?
            }
        }

        variables.push(Statement::ImportVar {
            name,
            value: Box::new(value),
        })
    }

    Ok(variables)
}

pub fn parse_import<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    lexer.next().transpose()?;

    let path = parse_string(
        source.as_ref(),
        lexer,
        "path for modules must be in string format",
        PATH_MSG,
    )?;

    let name = parse_identifier(
        source.as_ref(),
        lexer,
        "module name must be a valid identifier",
        IDENT_MSG,
    )?;

    let Statement::Address(address) = parse_simple_address(source.as_ref(), lexer, ADDRESS_HELP, ADDRESS_MSG)? else {
        unreachable!();
    };

    expect(
        Kind::LBrace,
        lexer,
        source.as_ref(),
        "modules must have a variable declaration block",
        LBRACE_MSG,
    )?;

    let variables = parse_import_vars(source.as_ref(), lexer)?;

    expect(
        Kind::RBrace,
        lexer,
        source.as_ref(),
        "unclosed module declaration block. you most likely forgot a `}` [RIGHT_CURLY]",
        RBRACE_MSG,
    )?;

    Ok(Statement::Import {
        name,
        path,
        address,
        variables,
    })
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_parse_import_with_variables() {
        let input = r#"
            import "./path.aya" module_name &[$fefe] {
                variable_a: $C0D3,
                variable_b: [!other_variable],
                variable_c: [other_module.variable],
                variable_d: !other_variable,
            }
        "#;
        let result = crate::parser::parse(input).unwrap();
        insta::assert_debug_snapshot!(result);
    }
}
