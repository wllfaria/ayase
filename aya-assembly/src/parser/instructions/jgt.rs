use crate::lexer::{Kind, Lexer, TransposeRef};
use crate::parser::ast::{Instruction, Statement};
use crate::parser::common::{expect, parse_hex_lit, parse_keyword, parse_register, parse_variable};
use crate::parser::error::{
    ADDRESS_HELP, ADDRESS_MSG, BRACKETED_EXPR_HELP, BRACKETED_EXPR_MSG, COMMA_MSG, HEX_LIT_HELP, HEX_LIT_MSG, VAR_HELP,
    VAR_MSG,
};
use crate::parser::expressions::{parse_address_expr, parse_literal_expr};
use crate::parser::Result;
use crate::utils::{unexpected_eof, unexpected_token};

pub fn parse_jgt<S: AsRef<str>>(source: S, lexer: &mut Lexer) -> Result<Statement> {
    parse_keyword(source.as_ref(), lexer, Kind::Jgt)?;

    let lhs = parse_address_expr(source.as_ref(), lexer, ADDRESS_HELP, ADDRESS_MSG)?;

    expect(
        Kind::Comma,
        lexer,
        source.as_ref(),
        "missing a comma after left side of instruction",
        COMMA_MSG,
    )?;

    let Ok(Some(token)) = lexer.peek().transpose() else {
        let Err(err) = lexer.next().transpose() else {
            return unexpected_eof(source.as_ref(), "unterminated import statement");
        };
        return Err(err);
    };

    let kind = token.kind;
    let rhs = match kind {
        Kind::Ident => Statement::Register(parse_register(source.as_ref(), lexer)?),
        Kind::HexNumber => Statement::HexLiteral(parse_hex_lit(source.as_ref(), lexer, HEX_LIT_HELP, HEX_LIT_MSG)?),
        Kind::Bang => Statement::Var(parse_variable(source.as_ref(), lexer, VAR_HELP, VAR_MSG)?),
        Kind::LBracket => parse_literal_expr(source.as_ref(), lexer, BRACKETED_EXPR_HELP, BRACKETED_EXPR_MSG)?,
        _ => return unexpected_token(source.as_ref(), token),
    };

    match kind {
        Kind::Ident => Ok(Instruction::JgtReg(lhs, rhs).into()),
        Kind::HexNumber => Ok(Instruction::JgtLit(lhs, rhs).into()),
        Kind::Bang => Ok(Instruction::JgtLit(lhs, rhs).into()),
        Kind::LBracket => Ok(Instruction::JgtLit(lhs, rhs).into()),
        _ => unreachable!(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_instruction(input: &str) -> Statement {
        let mut lexer = Lexer::new(input);
        parse_jgt(input, &mut lexer).unwrap()
    }

    #[test]
    fn test_jgt_reg() {
        let input = "jgt &[$c0d3], r2";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_jgt_reg_expr() {
        let input = "jgt &[$c0d3 + r2], r2";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_jgt_lit() {
        let input = "jgt &[$c0d3], $0303";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_jgt_lit_var() {
        let input = "jgt &[$c0d3], !var";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_jgt_lit_expr() {
        let input = "jgt &[$c0d3], [$0303 + r2]";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_jgt_lit_expr_both() {
        let input = "jgt &[$c0d3 + r2], [$0303 + r2]";
        let result = run_instruction(input);
        insta::assert_debug_snapshot!(result);
    }
}
