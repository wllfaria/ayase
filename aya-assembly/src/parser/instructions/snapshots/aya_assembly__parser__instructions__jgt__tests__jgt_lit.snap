---
source: aya-assembly/src/parser/instructions/jgt.rs
expression: result
---
Instruction(
    JgtLit(
        Address(
            ByteOffset {
                start: 7,
                end: 11,
            },
        ),
        HexLiteral(
            ByteOffset {
                start: 15,
                end: 19,
            },
        ),
    ),
)
